### PR TITLE
Remove implicit environ from storage

### PR DESCRIPTION
**Pull request by dhermes**
Has #987 as diffbase.

@tseaver The 2nd commit in this PR is all over the place. I can try to re-work if it seems too daunting.

It started out as just deleting `_implicit_environ` and then dealing with all the dependencies. This included
- ditching `gcloud.get_default_project()`
- axing a lot of the implicit test utils
- swapping the `_client_or_connection` for `_require_client` (and moving `Bucket._require_client` to `_PropertyMixin`)
- updating uses of `connection` (from `_client_or_connection`) to be `client.connection`
- adding `client` properties to `ACL` and `_PropertyMixin` methods
- removing the `storage._require_connection` method
- removing the concept of default bucket and making `bucket` required in `Blob` constructor
- updating `Bucket.create` so that both the project and connection come from the `client`
- removing lots of "implicit" tests that had explicit equivalents

### CODE REVIEW

**Comment by tseaver** [comment]
@dhermes Do you plan to rebase here after #987 lands?

**Comment by dhermes** [comment]
@tseaver Rebased on top of #987

File: gcloud/_testing.py [hunk 1]
```diff
@@ -35,11 +35,6 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
 
 
-def _monkey_defaults(*args, **kwargs):
-    mock_defaults = _DefaultsContainer(*args, **kwargs)
-    return _Monkey(_helpers, _DEFAULTS=mock_defaults)
-
-
 def _setup_defaults(test_case, *args, **kwargs):
     test_case._replaced_defaults = _helpers._DEFAULTS
     _helpers._DEFAULTS = _DefaultsContainer(*args, **kwargs)
```
Comment (tseaver) [inline gcloud/_testing.py]: I don't see a PR which removes the `_testing` module altogether (along with `_DefaultsContainer`).

File: gcloud/storage/acl.py [hunk 2]
```diff
@@ -351,23 +349,24 @@ def get_entities(self):
...
-        :returns: The connection determined from the ``client`` or environment.
+        :rtype: :class:`gcloud.storage.client.Client`
+        :returns: The client passed in or the currently bound client.
         """
         if client is None:
-            return _require_connection()
-        else:
-            return client.connection
+            client = self.client
+        return client
```
Comment (tseaver) [inline gcloud/storage/acl.py]: Seems like it would be good not to repeat this between `ACL` and `_PropertyMixin`.

File: gcloud/storage/test___init__.py [hunk 3]
```diff
@@ -1,155 +0,0 @@
...
-
-        with _Monkey(storage, set_default_bucket=call_set_bucket,
-                     set_default_connection=call_set_connection,
-                     set_default_project=call_set_project):
-            self._callFUT(bucket=BUCKET, project=PROJECT,
-                          connection=CONNECTION)
-
-        self.assertEqual(SET_PROJECT_CALLED, [PROJECT])
-        self.assertEqual(SET_CONNECTION_CALLED, [CONNECTION])
-        self.assertEqual(SET_BUCKET_CALLED, [BUCKET])
```
Comment (tseaver) [inline gcloud/storage/test___init__.py]: Yay!

File: gcloud/_testing.py (hunk 1 above)
Comment (dhermes) [inline gcloud/_testing.py]: We don't want to do that since we define `_Monkey` here.

File: gcloud/storage/acl.py (hunk 2 above)
Comment (dhermes) [inline gcloud/storage/acl.py]: I agree. Wasn't sure what to do about it, didn't want to make the classes too close cousins just yet. Shall I file a bug?

Also, `ACL` still needs a lot of TLC (actual design) from us. I think you'll need to tackle it as part of BigQuery too.

File: gcloud/storage/test_batch.py [hunk 4]
```diff
@@ -362,6 +354,16 @@ def test_finish_nonempty_non_multipart_response(self):
...
 
+    def test_current(self):
+        from gcloud.storage.batch import _BATCHES
+        klass = self._getTargetClass()
+        batch_top = object()
+        self.assertEqual(list(_BATCHES), [])
+        _BATCHES.push(batch_top)
+        self.assertTrue(klass.current() is batch_top)
+        _BATCHES.pop()
+        self.assertEqual(list(_BATCHES), [])
```
Comment (tseaver) [inline gcloud/storage/test_batch.py]: This test seems unrelated to the PR (a merge / rebase artefact?).

Given that it belongs here, it should probably test that `klass.current()` returns `None` before and after the `push()`/`pop()`.

Comment (dhermes) [inline gcloud/storage/test_batch.py]: Was put in to keep test coverage at 100% (some code path that used to call `.current()` is getting axed here). Ends up being irrelevant in #990 when the local stack is used on the client.

File: gcloud/_testing.py (hunk 1 above)
Comment (tseaver) [inline gcloud/_testing.py]: OIC.  I misfired on the `_setup_defaults` beneath it, which seems like it should go away.

Comment (dhermes) [inline gcloud/_testing.py]: Yup. Those go away in #991

**Comment by tseaver** [comment]
LGTM
